scripts/dataconnect-test/tests.ts

```diff
@@ -28,45 +63,128 @@ async function list() {
     { FIREBASE_CLI_EXPERIMENTS: "dataconnect" },
   );
 }
 
-async function deploy() {
-  return await cli.exec(
-    "deploy",
-    FIREBASE_PROJECT,
-    ["--only", "dataconnect", "--force"],
-    __dirname,
-    /** quiet=*/ false,
-    { FIREBASE_CLI_EXPERIMENTS: "dataconnect" },
-  );
+async function deploy(force: boolean) {
+  const args = ["--only", "dataconnect"];
+  if (force) {
+    args.push("--force");
+  }
+  if (FIREBASE_DEBUG) {
+    args.push("--debug");
+  }
+  return await cli.exec("deploy", FIREBASE_PROJECT, args, __dirname, /** quiet=*/ false, {
+    FIREBASE_CLI_EXPERIMENTS: "dataconnect",
+  });
+}
+
+function toPath(p: string) {
+  return path.join(__dirname, p);
+}
+
+function getRandomString(length: number): string {
+  const SUFFIX_CHAR_SET = "abcdefghijklmnopqrstuvwxyz0123456789";
+  let result = "";
+  for (let i = 0; i < length; i++) {
+    result += SUFFIX_CHAR_SET.charAt(Math.floor(Math.random() * SUFFIX_CHAR_SET.length));
+  }
+  return result;
+}
+
+// Each test run should use a random serviceId and databaseId.
+function newTestRun(): { serviceId: string; databaseId: string } {
+  const serviceId = `cli-e2e-service-${getRandomString(6)}`;
+  const databaseId = `cli-e2e-database-${getRandomString(6)}`;
+
+  const dataconnectYamlTemplate = fs.readFileSync(toPath("templates/dataconnect.yaml")).toString();
+  const connectorYamlTemplate = fs.readFileSync(toPath("templates/connector.yaml")).toString();
+  const subbedDataconnectYaml = dataconnectYamlTemplate
+    .replace("__serviceId__", serviceId)
+    .replace("__databaseId__", databaseId);
+  if (!fs.existsSync(toPath("fdc-test"))) {
+    fs.mkdirSync(toPath("fdc-test"));
+  }
+  if (!fs.existsSync(toPath("fdc-test/connector"))) {
+    fs.mkdirSync(toPath("fdc-test/connector"));
+  }
+  if (!fs.existsSync(toPath("fdc-test/schema"))) {
+    fs.mkdirSync(toPath("fdc-test/schema"));
+  }
+  fs.writeFileSync(toPath("fdc-test/dataconnect.yaml"), subbedDataconnectYaml, {
+    mode: 420 /* 0o644 */,
+  });
+  fs.writeFileSync(toPath("fdc-test/connector/connector.yaml"), connectorYamlTemplate, {
+    mode: 420 /* 0o644 */,
+  });
+  return { serviceId, databaseId };
+}
+
+function prepareStep(step: Step) {
+  fs.writeFileSync(toPath("fdc-test/schema/schema.gql"), step.schemaGQL, { mode: 420 /* 0o644 */ });
+  fs.writeFileSync(toPath("fdc-test/connector/connector.gql"), step.connectorGQL, {
+    mode: 420 /* 0o644 */,
+  });
 }
 
 describe("firebase deploy", () => {
-  before(() => {
+  let serviceId: string;
+  let databaseId: string;
+
+  beforeEach(async function (this) {
+    this.timeout(10000);
     expect(FIREBASE_PROJECT).not.to.equal("", "No FBTOOLS_TARGET_PROJECT env var set.");
+    const info = newTestRun();
+    serviceId = info.serviceId;
+    databaseId = info.databaseId;
+    await requireAuth({});
+  });
+
+  afterEach(async function (this) {
+    this.timeout(10000);
+    fs.rmSync(toPath("fdc-test"), { recursive: true, force: true });
```

Comment:
Extract a constant or variable for `toPath("fdc-test")`?

See it in many places